plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1371,6 +1378,144 @@ \subsubsection{Comparison of the typed CK machine and the untyped CEK machine}
+    \verb|0000010 0000000 0000000| $\rightarrow$ \verb|0000000 0000000 0000010|
+  \item Add list constructor tags: \\
+    \verb|0000000 0000000 0000010| $\rightarrow$ \verb|10000000 10000000 00000010|
+\end{enumerate}
+
+\subsection{Constants}
+
+Constants are encoded as a combination of the kind of value that is serialized and the
+value itself. Constants use 4 bits to encode the constructors, so they allow for a
+maximum of 16 constructors, of which 6 are currently used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|l|c|l|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Encoding} \\
+  \hline
+  ByteString & 1 & Variable length \rule{0mm}{4mm} \\[\sep]
+  Integer & 0 & ZigZag + Variable length \\[\sep]
+  String & 2 & UTF-8 \\[\sep]
+  Char & 3 & UTF-8 \\[\sep]
+  Unit & 4 & Empty \\[\sep]
+  Bool & 5 & `1' is True, `0' is False \\[\sep]
+  \hline
+\end{tabular}
+\captionof{figure}{Constant kinds}
+\label{fig:serialisation-constants}
+\end{minipage}
+
+\subsection{Untyped terms}
+
+Terms are encoded using 4 bits, which allows a total of 16 terms, of which 8 are
+currently used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|l|c|l|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Arguments} \\
+  \hline
+  Variable & 0 & name \rule{0mm}{4mm} \\[\sep]
+  Delay & 1 & term \\[\sep]
+  Lambda abstraction & 2 & name, term \\[\sep]
+  Application & 3 & term, term \\[\sep]
+  Constant & 4 & constant \\[\sep]
+  Force & 5 & term \\[\sep]
+  Error & 6 & term \\[\sep]
+  Builtin & 7 & builtin \\[\sep]
+  \hline
+\end{tabular}
+\captionof{figure}{Untyped terms}
+\label{fig:serialisation-terms}
+\end{minipage}
+
+\vspace{1cm}
+
+\noindent Builtins use 5 bits for their definition, allowing for a maximum of 32 builtin
+functions of which 25 are used.
+
+\vspace{1cm}
+
+\begin{minipage}{\linewidth}
+\centering
+\begin{tabular}{|c|c|c|c|c|c|}
+  \hline
+  \Strut
+  \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} \\
+  \hline
+   AddInteger & 0 & GreaterThanEqInteger & 8 & QuotientInteger & 17 \rule{0mm}{4mm} \\[\sep]
+   SubtractInteger & 1 & EqInteger & 9 & ModInteger & 18 \\[\sep]
+   MultiplyInteger & 2 & Concatenate & 10 & LtByteString & 19 \\[\sep]
+   DivideInteger & 3 & TakeByteString & 11 & GtByteString & 20 \\[\sep]
+   RemainderInteger & 4 & DropByteString & 12 & IfThenElse & 21 \\[\sep]
+   LessThanInteger & 5 & SHA2 & 13 & CharToString & 22 \\[\sep]
+   LessThanEqInteger & 6 & SHA3 & 14 & Append & 23 \\[\sep]
+   GreaterThanInteger & 7 & VerifySignature & 15 & Trace & 24 \\[\sep]
+   & & EqByteString & 16 & & \\[\sep]
+   \hline
+\end{tabular}
+\captionof{figure}{Builtin tags}
+\label{fig:serialisation-builtins}
+\end{minipage}
+
+\vspace{1cm}
+
+\noindent Encoding primtive types:
+
+\begin{itemize}
+  \item Unsigned \texttt{Integer}s and \texttt{ByteString}s are encoded using the
```

Comment:
Very good catch. I was secretly hoping I don't have to talk about pre-aligned data structures. I think, at some point, it would be *very* useful to have an example encoding of some program that covers all Plutus Core data types, together with a test to check that the encoding and example are in sync. This would prevent changes to GHC Generic deriving mechanism affecting us. @michaelpj @kwxm, should I create an issue for it?

Reply:
I have update the specification with the encoding for `ByteString`, `Char` and `String`.